### ISSUE TITLE
Buildingplan update plan

### ISSUE THREAD

**Issue by myk002** [DFHack/dfhack#2883]
Background
------------------

`bulidingplan` is one of DFHack's most critical tools. It allows a player (or another DFHack tool) to place buildings before the items required to build that building are available. The construction jobs remain in a suspended state until `buildingplan` has attached appropriate items. Then the construction job is unsuspended and the game assigns a dwarf to complete the job as normal. This gives the player the flexibility to design their fort when they want to, and dwarves will fulfill the player's designs as items become available.

`buildingplan` also allows the player to set filters on which items will be considered for attachment. They can select any number of specific materials or material classes (e.g. "stone"). In previous versions, a player would select a building type for building in the UI, then select the item index for the required building element, then select the `M` hotkey for setting the material filter. A screen would come up where they could select the filter configuration they want.

These filters were also applied when other tools create buildings and register them with `buildingplan`. For example, a common workflow would be to go through the material filter selection process for a building, but not actually build the building. Then the player would use `quickfort` or `automaterial` to designate the building patterns, and `buildingplan` would fulfill the plans with the saved filters.

This document describes how `buildingplan` might be improved for better usability and how it can be transformed to fit in with the new DF UI.

Feature set
-----------------

`buildingplan` should provide the following features:

- Allow players to place all building types regardless of building material availability
- Allow players to choose specific items to attach when items are available
- Protect the planned buildings from being accidentally unsuspended (which would cancel their construction jobs)
- Allow the player to configure filters for item selection
- Optionally generate manager orders for needed materials that match the filters
- Give players information about planned buildings, including which items are needed and the building's position in the queue
- Allow the player to boost the priority of a planned building so it gets access to materials first
- Provide the player with a summary of current state and pending actions

Orders generation might be a difficult problem to solve, since it will require figuring out the mapping from a filter to an item, and then cross-referencing that with available reagents. If we can provide this feature, though, we can turn down the minimally-configurable orders generation in `quickfort` and handle orders completely from `buildingplan`. As we implement this plan, orders integration will likely come last.

Architecture
------------------

Buildingplan will have four main components:

- Core planned building management service (C++ plugin, `plugins/buildingplan.cpp`)
- Building planner overlay (Lua overlay widget, `plugins/lua/buildingplan.lua`)
- Planned building status panel (Lua overlay widget, `plugins/lua/buildingplan.lua`)
- Configuration tool (Lua, `scripts/gui/buildingplan.lua`)

#### Core

The core service will provide an API for registering planned buildings and will run periodically to scan for available items that match the associated filters, attach the items to the planned buildings, and unsuspend the planned buildings when all items are attached. It will also provide APIs for setting options and retrieving state for display by the other components.

The core is also responsible for persisting and reloading state. Some user preference options will be global and will be shared across forts. State on planned buildings and building filters will be persisted and restored with the relevant savegame.

Note that the core plugin should not be disabled since it provides services to other tools (primarily `quickfort`). If a player does not want to use the buildingplan planning interface, the planner overlay can be disabled in `gui/control-panel` under the "Overlays" tab.

#### Planner overlay

The planner overlay replaces the vanilla building placement workflow. When the player has selected a building type to build, the overlay covers the part of the vanilla panel that has the "Use closest material", "Select specific material", and "Keep placing after placement" widgets. In that space, the overlay instead shows the materials that the building will require and offers widgets for configuring how those materials will be provided/chosen.

Mock-up screenshot:
![image](https://user-images.githubusercontent.com/977482/218297484-e4626113-bbea-45ba-8656-dce4a255ba0f.png)

#### Status overlay

The status overlay appears in empty space the "info sheet" when a planned building is selected and gives the player information about the items required for the planned building and the ability to modify configuration, including bumping the priority of the viewed building.

The status sheet will also warn if there are no items available and no active jobs/manager orders for producing needed items.

Mock-up screenshot:
![image](https://user-images.githubusercontent.com/977482/218297530-06d4f427-522d-403f-abfd-0a57ac3c9834.png)

#### Configuration tool

The `gui/buildingplan` config tool will allow the player to set options and see aggregated or per-building metrics. This will be the only interface for all configuration, including per-building filters, when the player intends to use `quickfort`, `automaterial`, or other tools to perform building placement.

Use cases
---------------

**A) Player wants to place buildings manually, but with the least amount of toil.**

This is likely to be the most common case.

1. player selects the desired building type from the vanilla build menu
2. player clicks on building direction options (if applicable) then on the map to place the building
3. building mode stays active as if the vanilla checkbox for multi-place were set
4. player drags or pans the screen and places more copies of the building as required
5. player hits Esc or right clicks to exit build mode

Each building placement will register the building with the buildingplan API. When buildingplan does its next scan, it will attach existing items to the building or, if no items are available, will generate manager orders to produce usable items from available materials.

**B) Player cares about which items are chosen**

This is a common case for `quickfort` users, where building placement is handled for them but they want, for example, all their walls to be made out of blocks and to be the same color.

1. player opens `gui/buildingplan`, either directly or from a button on the planner overlay or from another tool UI like `gui/quickfort`. If the building placement screen is active, `gui/buildingplan` opens directly to the configuration screen for the selected building.
2. player configures the filters as desired (this part will need design and discussion for exactly how we want to present the filter options)
3. player closes `gui/buildingplan` and continues with their workflow for building placement

**C) Player needs a specific planned building to get built ASAP**

1. player clicks on the planned building
2. player clicks on the "make top priority" button or hits the hotkey

item match tasks for this building are then moved to the front of `buildingplan`'s internal queue structures.

Note that there is no easy way to select a group of buildings. We'd need an interface like `gui/blueprint` where the player could box select and all selected buildings could be prioritized. If this is a desired feature (and I could see it becoming so -- e.g. "need to build this planned wall now before the invaders get here"), we could launch such a selection interface from the status overlay itself. This functionality might also fit in with a potential future `gui/prioritize` tool.

**Comment by myk002** [comment]
Planner overlay UX
----------------------------

The planner overlay needs to take up minimal screen space while communicating a lot of information and providing access to relevant configuration.

to summarize, for each item that the building requires, the overlay could show:
- item name
- number required
- number of items available that match the currently set filter
- the number of items that would be available if there were no filter
- button for choosing a specific item (disabled if no available items match current filter)
- button for setting the filter

clicking on the filter button could launch `gui/buildingplan` and open the page for the selected item for the current building type. Depending on the eventual design of `gui/buildingplan`, maybe there could just be one page for all items for a building type, and we'd only need one button for configuration instead of one button per item.

Filter options
--------------------

What kind of things will a player want to filter on? Here is a list of options in the previous version, plus popular requests:

- specific material (e.g. "marble") O(100)
- material type (e.g. "stone", "wood", "glass") O(10)
- color (which may not be feasible anymore now that there are so many colors. it might be just as good to specify the material and show the color next to the material) O(?)
- whether the material is fire-proof/magma-proof O(2)

**Comment by myk002** [comment]
@TaxiService 

From u/GingerCritter on [reddit](https://www.reddit.com/r/dwarffortress/comments/12zob2m/comment/jhtuwgu/?utm_source=share&utm_medium=web2x&context=3):

I still use the vanilla building interface because it gives you a visual preview of what the materials look like when you select them. If the DFHack building blueprint functionality was a little more streamlined and visual then I would use it.

**Comment by TaxiService** [comment]
we can definitely work on it being more streamlined... Idk about more "visual" 🤔 
I'll start considering including some sort of preview of the item/material somewhere... like in vanilla.

**Comment by myk002** [comment]
another option might be to have just the sprite for the currently selected item appear in a fixed space

**Comment by TaxiService** [comment]
yeah. It's a good compromise. A box with a preview of the selected item.

It doesn't let the player compare all the on-screen list elements simultaneously like a vanilla list, but at least it's _something!_
